[CSS Tree Counting Functions] Parsing and serialization

weinig opened the pull request:
#### 879adcb9ef860cc34bb757be45b8132efe361b0f
<pre>
[CSS Tree Counting Functions] Parsing and serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=291270">https://bugs.webkit.org/show_bug.cgi?id=291270</a>

Reviewed by Darin Adler.

Adds support for the CSS Values and Units 5 tree counting CSS functions,
sibling-count() and sibling-index() (<a href="https://drafts.csswg.org/css-values-5/#tree-counting)">https://drafts.csswg.org/css-values-5/#tree-counting)</a>

The new functionality is off by default behind the experimental feature
settings flag: CSSTreeCountingFunctionsEnabled.

Still needed is support for invalidating styles that use the tree-counting
functions when their values change due to sibling count changes.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-complex-unresolved-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-in-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/sibling-function-container-query-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/sibling-function-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/tree-scoped-sibling-function-expected.txt:
    - Updated test results.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add new settings flag.

* Source/WebCore/css/CSSValueKeywords.in:
    - Add new keywords.

* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
    - Hook up parsing and evaluation of the new functions. Evaluation
      is delegated to Style::BuilderState, as it has the Element needed
      to compute the values.

* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
    - Add cssTreeCountingFunctionsEnabled bit to allow the new functionality
      to be disabled at parse time.

* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
    - Implement evaluation of the functions using normal tree traversal.

Canonical link: <a href="https://commits.webkit.org/293468@main">https://commits.webkit.org/293468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/863cabddaece6ab58a139cee6191ff55f38267c4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99008 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18646 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104137 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49601 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101053 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27095 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75372 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32498 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102012 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14400 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89418 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55733 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14192 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7392 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48977 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91698 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84122 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7468 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106503 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97638 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26105 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19032 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84335 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26480 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85614 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83838 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21257 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28500 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6170 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19842 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26058 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31244 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121254 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25878 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33902 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29198 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27452 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,5 +332,31 @@ Ref<Calculation::RandomKeyMap> BuilderState::randomKeyMap(bool perElement) const
+    ASSERT(element());
+
+    unsigned count = 1;
+    for (const auto* sibling = ElementTraversal::previousSibling(*element()); sibling; sibling = ElementTraversal::previousSibling(*sibling))
```
Better as a `while` loop?

darinadler commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,5 +332,31 @@ Ref<Calculation::RandomKeyMap> BuilderState::randomKeyMap(bool perElement) const
+    unsigned count = 1;
+    for (const auto* sibling = ElementTraversal::previousSibling(*element()); sibling; sibling = ElementTraversal::previousSibling(*sibling))
+        ++count;
+    for (const auto* sibling = ElementTraversal::nextSibling(*element()); sibling; sibling = ElementTraversal::nextSibling(*sibling))
```
Better as a `while` loop?

darinadler commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,5 +332,31 @@ Ref<Calculation::RandomKeyMap> BuilderState::randomKeyMap(bool perElement) const
+    ASSERT(element());
+
+    unsigned count = 1;
+    for (const auto* sibling = ElementTraversal::previousSibling(*element()); sibling; sibling = ElementTraversal::previousSibling(*sibling))
```
Better as a `while` loop?

darinadler reviewed (approved):
Do you understand why the still failing tests are failing?

nt1m commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,5 +332,31 @@ Ref<Calculation::RandomKeyMap> BuilderState::randomKeyMap(bool perElement) const
+
+    ASSERT(element());
+
+    unsigned count = 1;
```
More of a spec question, but I find it odd that sibling-count() is self-including

nt1m replied:
https://github.com/w3c/csswg-drafts/issues/12062

weinig committed:
[CSS Tree Counting Functions] Parsing and serialization
https://bugs.webkit.org/show_bug.cgi?id=291270

Reviewed by Darin Adler.

Adds support for the CSS Values and Units 5 tree counting CSS functions,
sibling-count() and sibling-index() (https://drafts.csswg.org/css-values-5/#tree-counting)

The new functionality is off by default behind the experimental feature
settings flag: CSSTreeCountingFunctionsEnabled.

Still needed is support for invalidating styles that use the tree-counting
functions when their values change due to sibling count changes.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-complex-unresolved-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-in-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/sibling-function-container-query-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/sibling-function-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/tree-scoped-sibling-function-expected.txt:
    - Updated test results.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add new settings flag.

* Source/WebCore/css/CSSValueKeywords.in:
    - Add new keywords.

* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
    - Hook up parsing and evaluation of the new functions. Evaluation
      is delegated to Style::BuilderState, as it has the Element needed
      to compute the values.

* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
    - Add cssTreeCountingFunctionsEnabled bit to allow the new functionality
      to be disabled at parse time.

* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
    - Implement evaluation of the functions using normal tree traversal.

Canonical link: https://commits.webkit.org/293468@main
